libs/ballot-interpreter/README.md

```diff
@@ -96,6 +96,114 @@ To update the saved results for a benchmark, run:
+effectively begin at (1, 1).
+
+#### Timing Mark Detection and Error Handling
+
+Candidate timing marks are detected using a contours algorithm (similar to
+openCV's
+[`findContours`](https://docs.opencv.org/3.4/d4/d73/tutorial_py_contours_begin.html)
+function). This returns all the contiguous shapes as rectangle bounding boxes.
+We then filter these by plausibility based on size and position. We then find
+the top/bottom/left/right timing mark edges independently. This is done by
+splitting the candidate timing mark rectangles into top/bottom halves and
+left/right halves, then searching within each of those four halves for the line
+that intersects the most rectangles and is in a plausible position and angle.
+
+If each edge is found, we now have what the code refers to as "partial" timing
+marks. We then begin inferring any timing marks that we may have missed. We do
+this by examining the spacing between each ordered pair of timing marks and, if
+another would fit between them, we add it. When using bottom timing marks to
+encode the ballot metadata, we infer a lot of bottom timing marks because we
+expect many or even most of them to be missing. Once the inference is complete,
+we now have what the code refers to as "complete" timing marks. This is the data
+structure we use to determine the ballot layout and locate the bubbles.
+
+Handling errors in timing mark detection requires a balance between being too
+restrictive and rejecting too many ballots that could have been successfully
+interpreted and being too permissive and interpreting ballots incorrectly.
+Because the outcome of the former is not so bad–the voter may simply scan the
+ballot again–we err on the side of caution. We have a few strategies for
+handling errors:
+
+1. **Allow for some missing timing marks.** We infer missing timing marks based
+   on the spacing between known timing marks. This is especially important for
+   bottom timing marks, which may be used to encode the ballot metadata. We
+   can't allow too many missing timing marks, however, because the inference
+   process is not perfect and may introduce errors.
+2. **Require small rotation and/or skew if we've inferred any timing marks.** If
+   we didn't infer any timing marks we can be more lenient, but if we did we
+   need to be more strict about rotation and skew because they can cause the
+   inferred timing marks to be incorrect. Note that this project uses
+   "rotation" to refer to an angle by which the entire ballot is rotated,
+   preserving the equal distance of points in the grid from each other. "Skew"
+   refers to a distortion of the grid, where the distance between points in the
+   grid is not preserved, and is caused by different parts of the ballot being
+   scanned at different speeds.
+3. **Require that the number of inferred timing marks on a side matches the
+   expected number.** And that the left/right and top/bottom marks have the same
+   count.
```

Comment:
> And that the left/right and top/bottom marks have the same count

I assume the one exception to top/bottom marks having the same count is if we're encoding metadata using the bottom timing marks, is that correct?

Reply:
Yes, that's correct. Do you think it's worth clarifying here?

Reply:
I think it's worth a quick mention!

Reply:
https://github.com/votingworks/vxsuite/pull/5205